game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/retreat/OffensiveGeneralRetreat.java

```diff
@@ -0,0 +1,129 @@
+package games.strategy.triplea.delegate.battle.steps.retreat;
+
+import static games.strategy.triplea.delegate.battle.BattleStepStrings.ATTACKER_WITHDRAW;
+
+import games.strategy.engine.data.GameData;
+import games.strategy.engine.data.Territory;
+import games.strategy.engine.data.Unit;
+import games.strategy.engine.delegate.IDelegateBridge;
+import games.strategy.triplea.Properties;
+import games.strategy.triplea.delegate.ExecutionStack;
+import games.strategy.triplea.delegate.Matches;
+import games.strategy.triplea.delegate.battle.BattleActions;
+import games.strategy.triplea.delegate.battle.BattleState;
+import games.strategy.triplea.delegate.battle.MustFightBattle.RetreatType;
+import games.strategy.triplea.delegate.battle.steps.BattleStep;
+import games.strategy.triplea.delegate.battle.steps.RetreatChecks;
+import java.util.Collection;
+import java.util.List;
+import java.util.Set;
+import javax.annotation.Nullable;
+import lombok.AllArgsConstructor;
+import lombok.Value;
+import org.triplea.java.collections.CollectionUtils;
+
+@AllArgsConstructor
+public class OffensiveGeneralRetreat implements BattleStep {
+
+  final BattleState battleState;
+
+  final BattleActions battleActions;
+
+  @Override
+  public List<String> getNames() {
+    if (canAttackerRetreat()
+        || canAttackerRetreatSeaPlanes()
+        || (battleState.isAmphibious()
+            && (canAttackerRetreatPartialAmphib() || canAttackerRetreatAmphibPlanes()))) {
+      return List.of(battleState.getAttacker().getName() + ATTACKER_WITHDRAW);
+    }
+    return List.of();
+  }
+
+  @Override
+  public Order getOrder() {
+    return Order.OFFENSIVE_GENERAL_RETREAT;
+  }
+
+  @Override
+  public void execute(final ExecutionStack stack, final IDelegateBridge bridge) {
+    retreatUnits(bridge);
```

Comment:
It's really odd for `retreatUnits` to be the implementation of this interface method, and also called externally directly. Can you speak to why that is @trevan ?

Reply:
The same logic (retreat) is called in other steps.  I either had the choice of creating a static helper class to store the logic or to allow the other steps to instantiate and call this method directly.

I would prefer that the other steps didn't need this logic and hope that I might be able to fix it but that will involve changing the logic of the steps and might even change the game play.  I'd like to make that change in a separate PR.

Reply:
Fair enough. Extracting logic for re-use would seemingly be okay. Passing an instance to a constructor can help avoid static coupling.